Crossplane pod crashes on claim deletion

### What happened?
When I try to delete claim crossplane pod falls into crashing loop spiting panic errors into log. The only way to break the look is to delete finalizers on claims that should be deleted and then manually cleanup XRs.

```
I0526 21:24:10.243020       1 leaderelection.go:248] attempting to acquire leader lease crossplane-system/crossplane-leader-election-core...
I0526 21:25:12.084740       1 leaderelection.go:258] successfully acquired lease crossplane-system/crossplane-leader-election-core
W0526 21:25:12.088719       1 warnings.go:70] This API is deprecated and is scheduled to be removed in a future release.
W0526 21:25:12.091059       1 warnings.go:70] This API is deprecated and is scheduled to be removed in a future release.
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x1951d97]

goroutine 2389 [running]:
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile.func1()
	sigs.k8s.io/controller-runtime@v0.14.6/pkg/internal/controller/controller.go:119 +0x1fa
panic({0x1e38860, 0x33b72e0})
	runtime/panic.go:884 +0x213
github.com/crossplane/crossplane/internal/controller/apiextensions/claim.(*Reconciler).Reconcile(0xc002061380, {0x2431d90, 0xc006992240}, {{{0xc004244e40, 0x21}, {0xc004244e10, 0x21}}})
	github.com/crossplane/crossplane/internal/controller/apiextensions/claim/reconciler.go:437 +0x1537
github.com/crossplane/crossplane-runtime/pkg/ratelimiter.(*Reconciler).Reconcile(0xc0003d9860, {0x2431d90, 0xc006992240}, {{{0xc004244e40?, 0x0?}, {0xc004244e10?, 0x40dfe7?}}})
	github.com/crossplane/crossplane-runtime@v0.20.0-rc.0.0.20230330081344-bc8be4cd89f3/pkg/ratelimiter/reconciler.go:54 +0x16b
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile(0x2431d90?, {0x2431d90?, 0xc006992240?}, {{{0xc004244e40?, 0x1d38d40?}, {0xc004244e10?, 0x0?}}})
	sigs.k8s.io/controller-runtime@v0.14.6/pkg/internal/controller/controller.go:122 +0xc8
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler(0xc00098a6e0, {0x2431ce8, 0xc0003d98b0}, {0x1ee0e80?, 0xc00641f8c0?})
	sigs.k8s.io/controller-runtime@v0.14.6/pkg/internal/controller/controller.go:323 +0x35f
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem(0xc00098a6e0, {0x2431ce8, 0xc0003d98b0})
	sigs.k8s.io/controller-runtime@v0.14.6/pkg/internal/controller/controller.go:274 +0x1d9
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2()
	sigs.k8s.io/controller-runtime@v0.14.6/pkg/internal/controller/controller.go:235 +0x85
created by sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2
	sigs.k8s.io/controller-runtime@v0.14.6/pkg/internal/controller/controller.go:231 +0x587
```


### How can we reproduce it?
Not sure, it occurs only on 1.12.1. Getting back to 1.11.0 returned everything back to normal.

### What environment did it happen in?
Crossplane version:  1.12.1
EKS version 1.25

--- a/internal/controller/apiextensions/claim/reconciler.go
+++ b/internal/controller/apiextensions/claim/reconciler.go
@@ -411,8 +411,12 @@ func (r *Reconciler) Reconcile(ctx context.Context, req reconcile.Request) (reco
 
 		cm.SetConditions(xpv1.Deleting())
 		if meta.WasCreated(cp) {
+			requiresForegroundDeletion := false
+			if cdp := cm.GetCompositeDeletePolicy(); cdp != nil && *cdp == xpv1.CompositeDeleteForeground {
+				requiresForegroundDeletion = true
+			}
 			if meta.WasDeleted(cp) {
-				if *cm.GetCompositeDeletePolicy() == xpv1.CompositeDeleteForeground {
+				if requiresForegroundDeletion {
 					log.Debug("Waiting for the Composite to finish deleting (foreground deletion)")
 					return reconcile.Result{Requeue: true}, nil
 				}
@@ -434,7 +438,7 @@ func (r *Reconciler) Reconcile(ctx context.Context, req reconcile.Request) (reco
 			}
 
 			do := &client.DeleteOptions{}
-			if *cm.GetCompositeDeletePolicy() == xpv1.CompositeDeleteForeground {
+			if requiresForegroundDeletion {
 				client.PropagationPolicy(metav1.DeletePropagationForeground).ApplyToDelete(do)
 			}
 			if err := r.client.Delete(ctx, cp, do); resource.IgnoreNotFound(err) != nil {
@@ -444,7 +448,7 @@ func (r *Reconciler) Reconcile(ctx context.Context, req reconcile.Request) (reco
 				cm.SetConditions(xpv1.ReconcileError(err))
 				return reconcile.Result{Requeue: true}, errors.Wrap(r.client.Status().Update(ctx, cm), errUpdateClaimStatus)
 			}
-			if *cm.GetCompositeDeletePolicy() == xpv1.CompositeDeleteForeground {
+			if requiresForegroundDeletion {
 				log.Debug("Requeue to wait for the Composite to finish deleting (foreground deletion)")
 				return reconcile.Result{Requeue: true}, nil
 			}

